src/renderers/shared/fiber/ReactFiberClassComponent.js

```diff
@@ -204,10 +205,17 @@ module.exports = function(
   function constructClassInstance(workInProgress : Fiber) : any {
     const ctor = workInProgress.type;
     const props = workInProgress.pendingProps;
-    const context = getMaskedContext(workInProgress);
+    const unmaskedContext = getUnmaskedContext(workInProgress);
+    const context = getMaskedContext(workInProgress, unmaskedContext);
     const instance = new ctor(props, context);
     adoptClassInstance(workInProgress, instance);
     checkClassInstance(workInProgress);
+
+    // Cache unmasked context so we can avoid recreating masked context unless necessary.
+    // ReactFiberContext usually updates this cache but can't for newly-created instances.
+    instance.__reactInternalMemoizedUnmaskedChildContext = unmaskedContext;
+    instance.__reactInternalMemoizedMaskedChildContext = context;
```

Comment:
Ideally we also shouldn't use `__reactInternal[...]` fields outside of `FiberContext` file. Not that it matters a lot but it’s hacky and I want to keep it contained.